README.md
# manifail

[![Build Status](https://travis-ci.org/tulos/manifail.png?branch=master)](https://travis-ci.org/tulos/manifail)

Handle failures and manage retries without callbacks!

The semantics are modeled closely to
[Failsafe](https://github.com/jhalterman/failsafe), which has a Clojure wrapper
called [Diehard](https://github.com/sunng87/diehard).

Manifail differs in that it tries to be composable and not rely on callbacks
for the failure handling logic. The drawback is that it's less precise and
carries more overhead than Failsafe. Also, circuit breakers are currently not
supported.

## Disclaimer

Use Manifail if you're feeling adventurous and/or like to write your code in a
direct callback-less style! Otherwise please consider using Failsafe directly
or through Diehard.

## Usage

`[tulos/manifail "0.4.0"]`

First, require the namespaces:

```clojure
(refer-clojure :exclude '(delay))
(require '[manifold.deferred :as d])
(require '[manifold.executor :as ex])
(use 'manifail)
```

### Retry building blocks

To create a retriable execution with Manifail you have to use four parts
working in concert:

1. A retry policy - sequence of milliseconds representing the delay of retries
2. A piece of code potentially having the `retry`/`abort`/`reset` markers
3. An executor to run the code and retry attempts on
4. `with-retry` macro or `with-retry*` function wrapping the code to be retried

It looks like this:

```clojure
(def unreliable-service-executor (Executors/newFixedThreadPool 1))

(ex/with-executor unreliable-service-executor
  (let [retry-delays [10 50 100]]
    (with-retries retry-delays
      (try (let [result (unreliable-service)]
             (when (:error result)
               (retry! result)))
             result)
           (catch UnrecoverableException e
             (abort! e))
           (catch SessionExpiredException e
             (authenticate!)
             (reset! retry-delays)))
```

The `with-retries` block above will return a deferred of its result. This
deferred will get fulfilled when either the original call or one of the three
retries completes. The execution will happen on the specified
`unreliable-service-executor`.

Given the above code, a retry will happen if:

* An exception is thrown which is not an `UnrecoverableException`
* The `result` has an `:error` key

and the resulting deferred will be completed when:

* A non `:error` response is returned
* An `UnrecoverableException` happens and the execution is aborted. In this
  case a `manifail.Aborted` exception is thrown with the cause set to `e`
* A `manifail.RetriesExceeded` exception is thrown when there are no more retry
  attempts.

In case a `SessionExpiredException` is thrown the whole execution cycle will
begin anew with the supplied sequence of `retry-delays`. In the example above
it's the same sequence as the one provided to `with-retries` originally.

### Retry policies

The simplest possible retry logic that will perform 3 retries with delays of
50, 100 and 150 ms:

```clojure
(with-retries [50 100 150]
  (unreliable-service))
```

The following code will retry the call 5 times with 50 ms delay:

```clojure
(with-retries (delay (retries 5) 50)
  (unreliable-service))
```

same, but with a backoff factor of 2.0:

```clojure
(with-retries (delay (retries 5) 50 {:backoff-factor 2.0})
  (unreliable-service))
```

with a random jitter of 25 ms:

```clojure
(with-retries (delay (retries 5) 50 {:backoff-factor 2.0, :jitter-ms 25})
  (unreliable-service))
```

with a random jitter scaling the delay up to 1/2:

```clojure
(with-retries (delay (retries 5) 50 {:backoff-factor 2.0, :jitter-factor 0.5})
  (unreliable-service))
```

with a total call and retry duration limited to 500 ms:

```clojure
(with-retries (-> (retries 5)
                  (delay 50 {:backoff-factor 2.0, :jitter-factor 0.5})
                  (limit-duration 500))
  (unreliable-service))
```

unlimited equally spaced out retries with duration limited to 500 ms:

```clojure
(with-retries (-> (forever) (delay 50) (limit-duration 500))
  (unreliable-service))
```

### Retry context

There are several dynamic bindings available in the retry block:

* `*retry-count*` - the current retry count
* `*elapsed-ms*` - milliseconds elapsed since entering the retry block
* `*last-result*` - last result/exception that caused the retry

### A complete example

Below I tried to map callbacks available in Failsafe retry policy to the code
that you could write in order to get equivalent behaviour.  We assume that
`call-some-service` blocks until completed:

```clojure
(->
  (with-retries (delay (retries 5) 50) ;; retry on any exception by default
    (println "-- on before execution")
    (when (> *retry-count* 0)
      (println "-- on before retry"))
    (try (let [result (call-some-service)]
           (when (> result 5) ;; abort condition
             (println "-- on failed attempt 1")
             (println "-- on before abort")
             (abort! result)) ;; set the value of Abort to `result`
           (when (= result ::bad) ;; retry condition
             (println "-- on failed attempt 2")
             (retry!))
           result)
         (catch OkException _ :ok) ;; do not retry on this exception
         (catch UnrecoverableException e
           (abort! e)) ;; set the cause of Abort to `e`
         (catch RecoverableException _
           (do-some-recovery)
           (reset! (delay (retries 5) 50)))
         (catch Throwable e
           (when-not (marker? e)
             (println "-- on failed attempt any"))
           (throw e))))
  (d/chain #(println "-- on complete" %))
  (d/catch manifail.Aborted #(println "-- on after abort" (unwrap %))
  (d/catch manifail.RetriesExceeded #(println "-- on retries exceeded" (unwrap %))
  (d/catch #(println "-- on failure" %)))
```

`call-some-service` might also be asynchronous and return something derefable.
In this case the code within `with-retries` can `chain` the logic:

```clojure
(with-retries (delay (retries 5) 50)
  (-> (call-some-service)
      (d/chain (fn [result] ...))))
```

## License

Copyright © 2016 Tulos Capital

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log

## [Unreleased]

## 0.4.0 - 2016-08-31
### Added
- `*last-result*/*retry-count*/*elapsed-ms*` dynamic variables available inside
  the retry block
- `unwrap` to get the value/cause of the marker
### Changed
- `retry!/abort!` accept an object as an argument in addition to a `Throwable` cause
- `retry!/abort!` markers produce `Throwable` exceptions instead of `ExceptionInfo`
- All special exceptions subclass `Throwable` instead of `RuntimeException`

## 0.3.0 - 2016-08-30
### Added
- `reset!` to reset the execution with a new delays sequence

## 0.2.0 - 2016-08-16
### Changed
- `abort!` and `retry!` can take a `Throwable` cause as their argument

## 0.1.0 - 2016-08-15
### Added
- Initial release: `with-retries`, `forever`, `retries`, `delay`,
  `limit-duration`, `limit-retries`

[Unreleased]: https://github.com/your-name/manifail/compare/0.4.0...HEAD


test/manifail_test.clj
(ns manifail-test
  (:require [clojure.test :refer :all]
            [manifold
             [deferred :as d]
             [stream :as s]
             [executor :as ex]]
            [manifail :as sut])
  (:import [java.util.concurrent Executors]))

(defn- ->seq [s] (vec s))

(deftest fixed-retry-times
  (is (= 1 (count (->seq (sut/retries 1)))))
  (is (= 5 (count (->seq (sut/retries 5))))))

(deftest limited-times
  (is (= 0 (-> 5 sut/retries (sut/limit-retries 0) ->seq count)))
  (is (= 1 (-> 5 sut/retries (sut/limit-retries 1) ->seq count)))
  (is (= 5 (-> 5 sut/retries (sut/limit-retries 6) ->seq count))))

(deftest limited-duration
  (is (= 0 (-> 1 sut/retries (sut/limit-duration 0) ->seq count)))

  (let [s (sut/limit-duration (sut/retries 100) 5)
        cnt (atom 0)]
    (d/loop [s' s]
      (Thread/sleep 1)
      (-> (or (first s') (d/success-deferred ::done))
          (d/chain
            (fn [result]
              (swap! cnt inc)
              (when-not (= ::done result) (d/recur (rest s')))))))
    (is (>= 5 @cnt))))

(defn around [expected-ms took-ms]
  (<= (- expected-ms 1) took-ms (+ expected-ms 1)))

(defn between [lower-ms upper-ms took-ms]
  (<= (- lower-ms 1) took-ms (+ upper-ms 1)))

(defn- realized-seq [s]
  s)

(deftest delayed
  (testing "simple delay"
    (is (= [50] (realized-seq (sut/delay (sut/retries 1) 50)))))

  (testing "backoff factor"
    (is (= [10 20 40] (realized-seq (sut/delay (sut/retries 3) 10 {:backoff-factor 2.0})))))

  (testing "backoff factor and max delay"
    (is (= [10 20 20]
           (realized-seq (sut/delay (sut/retries 3) 10
                                    {:backoff-factor 2.0, :max-delay-ms 20})))))

  (testing "jitter-ms"
    (let [[a b c] (realized-seq
                    (sut/delay (sut/retries 3) 20 {:jitter-ms 5}))]
      (is (around 20 a))
      (is (between 15 25 b))
      (is (between 10 30 c))))

  (testing "jitter-factor"
    (let [[a b c] (realized-seq
                    (sut/delay (sut/retries 3) 20 {:jitter-factor 0.5}))]
      (is (around 20 a))
      (is (between 10 30 b))
      (is (between 5 45 c)))))

(defn- aborted?
  ([r] (aborted? (constantly true) r))
  ([ex-pred r]
   (try @r
        false
        (catch manifail.Aborted e (ex-pred e)))))

(defn- retries-exceeded?
  ([r] (retries-exceeded? (constantly true) r))
  ([ex-pred r]
   (try @r
        false
        (catch manifail.RetriesExceeded e (ex-pred e)))))

(deftest with-retries
  (let [executed (atom 0)]

    (testing "executes once"
      (is
        (= ::done
           @(sut/with-retries* (sut/retries 1)
              #(do (swap! executed inc) ::done))))
      (is (= 1 @executed)))

    (reset! executed 0)

    (testing "executes several times, retries using value"
      (is
        (retries-exceeded? #(= (sut/unwrap %) sut/retry)
          (sut/with-retries* (sut/retries 2)
            #(do (swap! executed inc)
                 sut/retry))))
      (is (= 3 @executed)))

    (reset! executed 0)

    (testing "executes several times, retries using retry!"
      (is
        (retries-exceeded? #(nil? (sut/unwrap %))
          (sut/with-retries* (sut/retries 2)
            #(do (swap! executed inc)
                 (sut/retry!)))))
      (is (= 3 @executed)))

    (reset! executed 0)

    (testing "retries exceeded carries cause"
      (is
        (retries-exceeded? #(= (.getMessage (sut/unwrap %)) "boom3")
          (sut/with-retries* (sut/retries 2)
            #(do (swap! executed inc)
                 (sut/retry! (Exception. (str "boom" @executed)))))))
      (is (= 3 @executed)))

    (reset! executed 0)

    (testing "executes several times, retries using generic exception"
      (is
        (retries-exceeded? #(= (.getMessage (sut/unwrap %)) "boom")
          (sut/with-retries* (sut/retries 2)
            #(do (swap! executed inc)
                 (throw (Exception. "boom"))))))
      (is (= 3 @executed)))

    (reset! executed 0)

    (testing "aborts execution using return value"
      (is
        (aborted?
          (sut/with-retries* (sut/retries 3)
            #(do (swap! executed inc)
                 (if (= @executed 2)
                   sut/abort
                   sut/retry)))))
      (is (= 2 @executed)))

    (reset! executed 0)

    (testing "aborts execution after some time"
      (is
        (aborted?
          (sut/with-retries* (sut/retries 3)
            #(do (swap! executed inc)
                 (if (= @executed 2)
                   (sut/abort!)
                   (sut/retry!))))))
      (is (= 2 @executed)))

    (reset! executed 0)

    (testing "sets abort! cause"
      (is
        (aborted? #(= (.getMessage (sut/unwrap %)) "boom")
          (sut/with-retries* (sut/retries 3)
            #(sut/abort! (Exception. "boom")))))
      (is (= 0 @executed)))

    (reset! executed 0)

    (testing "sets abort! value"
      (is
        (aborted? #(= (sut/unwrap %) "boom")
          (sut/with-retries* (sut/retries 3)
            #(sut/abort! "boom"))))
      (is (= 0 @executed)))

    (reset! executed 0)

    (testing "limits duration"
      (is
        (retries-exceeded?
          (sut/with-retries* (sut/limit-duration (sut/retries 3) 50)
            #(do (swap! executed inc)
                 (Thread/sleep 30)
                 (sut/retry!)))))
      (is (= 2 @executed)))

    (reset! executed 0)

    (testing "limits duration with delay"
      (is
        (retries-exceeded?
          (sut/with-retries* (sut/limit-duration (sut/delay (sut/retries 3) 20) 50)
            #(do (swap! executed inc)
                 (Thread/sleep 30)
                 (sut/retry!)))))
      (is (= 2 @executed)))))

(deftest with-retries-reset
  (let [executed (atom 0)]

    (testing "resets"
      (is
        (retries-exceeded? #(= (sut/unwrap %) 5)
          (sut/with-retries* (sut/retries 2)
            #(do (swap! executed inc)
                 (if (= @executed 3)
                   (sut/reset! (sut/retries 1))
                   (sut/retry! @executed))))))
      ;; (1 execution + 2 retries initially) + (1 execution + 1 retry after reset)
      (is (= 5 @executed)))))

(deftest dynamic-vars
  (let [states (atom [])
        eps-ms 15, execution-ms 50, delay-ms 100]
    (is
      (retries-exceeded?
        (sut/with-retries* (repeat 2 delay-ms)
          #(do (swap! states conj {:elapsed sut/*elapsed-ms*
                                   :last-result sut/*last-result*
                                   :retrying sut/*retry-count*})
               (Thread/sleep execution-ms)
               (sut/retry! (count @states))))))

    (is (= 3 (count @states)))
    (let [{:keys [elapsed last-result retrying]} (first @states)]
      (is (<= elapsed eps-ms))
      (is (= retrying 0))
      (is (= last-result ::sut/none)))

    (let [{:keys [elapsed last-result retrying]} (second @states)]
      (is (<= elapsed (+ execution-ms delay-ms eps-ms)))
      (is (= retrying 1))
      (is (= last-result 1)))

    (let [{:keys [elapsed last-result retrying]} (nth @states 2)]
      (is (<= elapsed (+ (* 2 (+ execution-ms delay-ms eps-ms)))))
      (is (= retrying 2))
      (is (= last-result 2)))))


src/clj/manifail.clj
(ns manifail
  (:refer-clojure :exclude [delay reset!])
  (:require [manifold
             [deferred :as d]
             [executor :as ex]])
  (:import [clojure.lang ExceptionInfo]
           [manifail Aborted Retried RetriesExceeded Reset]))

(defn forever
  "A sequence of infinite retries with zero delay."
  [] (repeat 0))

(defn retries
  "Creates a sequence of `n` retries with zero delay."
  [n] (take n (forever)))

(defn- verify-factor [v min max name]
  (assert (and v (> v min) (<= v max))
          (format "%s factor must be between %s and %s!" name min max)))

(defn- delay-opts [delay-ms {:keys [max-delay-ms backoff-factor jitter-ms jitter-factor]}]
  (assert (not (and jitter-ms jitter-factor))
          "Cannot have both jitter-ms and jitter-factor specified!")
  (when max-delay-ms (assert (> max-delay-ms 0) "Max delay must be positive!"))
  (when jitter-ms (assert (> jitter-ms 0) "Jitter delay must be positive!"))
  (when jitter-factor (verify-factor jitter-factor 0.0 1.0 "Jitter"))
  (when backoff-factor (verify-factor backoff-factor 0.0 Double/MAX_VALUE "Backoff"))
  {:max-delay-ms (long (or max-delay-ms Long/MAX_VALUE))
   :backoff-factor (if backoff-factor (double backoff-factor) 1.0)
   :jitter-ms (if jitter-ms (long jitter-ms) 0)
   :jitter-factor (if jitter-factor (double jitter-factor) 0.0)})

(defn- compute-jitter [{:keys [jitter-ms jitter-factor]}]
  (let [rand-move #(- 1 (* 2 (rand)))]
    (cond
      (> jitter-ms 0) #(long (+ % (* jitter-ms (rand-move))))
      (> jitter-factor 0) #(long (* % (inc (* jitter-factor (rand-move)))))
      :else identity)))

(defn delay
  "Computes a delay for each retry in the `src` seq according to the `opts`.

  When no `opts` given, every delay will be the same `delay-ms`.

  Possible `opts`:
    * `max-delay-ms` - the maximum delay
    * `backoff-factor` - the factor to multiply each delay with,
      e.g. a factor of 2.0 and delay of 50ms will produce delays of
      50, 100, 200, 400, ...
    * `jitter-ms` - the amount of time to jitter each delay with,
      e.g. a jitter of 25ms and a delay of 50ms will result in delays of
      50, [25; 75], [0; 100], ... or concretely 50, 60, 45, 53, ...
    * `jitter-factor` - the factor between 0 and 1 to jitter each delay with,
      e.g. a factor of 0.5 and a delay of 50ms will result in delays of
      50, [25; 75], [12; 112], ... or concretely 50, 70, 100, 65, ..."
  ([src delay-ms] (delay src delay-ms {}))
  ([src delay-ms opts]
   (if (> delay-ms 0)
     (let [{:keys [max-delay-ms backoff-factor] :as opts} (delay-opts delay-ms opts)
           next-delay (volatile! delay-ms)
           jitter (compute-jitter opts)
           advance #(Math/max 0 (jitter (Math/min (long (* % backoff-factor)) max-delay-ms)))]
       (for [_ src]
         (let [d @next-delay]
           (vswap! next-delay advance)
           d)))
     src)))

(defn limit-retries
  "Limits the number of retries available through seq `src` to
  `max-retries` - a non-negative integer."
  [src max-retries]
  (let [max-retries (Math/max max-retries (int 0))]
    (if (> max-retries 0)
      (take max-retries src)
      [])))

(defn limit-duration
  "The retries available through seq `src` will be available to take for
  `max-duration-ms` milliseconds - a non-negative float. The countdown starts
  once `limit-duration` is called.

  The last delay will be the minimum of the remaining time and the delay
  received from `src`. E.g. if execution takes 30 ms and we have a source of
  constant 30 ms delays with duration limited to 140 ms, then:

    execution:                                   #1  #2  #3  #4
    execution takes:                             30  30  30  30
    time remaining before running delay:          -  80  20   -
    delay:                                        -  30  20   -
    time remaining when next execution started: 110  50   0   -"
  [src max-duration-ms]
  (let [max-duration-ms (Math/max max-duration-ms 0)
        elapsed-since #(- (System/currentTimeMillis) %)]
    (if (> max-duration-ms 0)
      (let [start-ms (System/currentTimeMillis)
            this (fn this [items]
                   (lazy-seq
                     (let [remaining (- max-duration-ms (elapsed-since start-ms))]
                       (when (and (seq items) (> remaining 0))
                         (cons (min (first items) remaining)
                               (this (rest items)))))))]
        (this src))
      [])))

(def ^{:doc "The return value to be used in case the execution needs to be
            retried"}
  retry ::retry)
(def ^{:doc "The return value to be used in case the execution needs to be
            aborted (stopped from being retried)"}
  abort ::abort)

(let [e (Aborted.)]
  (defn abort!
    "Throws an exception which short-circuits the execution.

    Should only be used inside the retriable code. No more retries will be
    performed after an abort."
    ([] (throw e))
    ([v]
     (throw (if (instance? Throwable v)
              (Aborted. ^Throwable v)
              (Aborted. ^Object v))))))

(let [e (Retried.)]
  (defn retry!
    "Short-circuits the execution for the next retry.

    Should only be used inside the retriable code."
    ([] (throw e))
    ([v]
     (throw (if (instance? Throwable v)
              (Retried. ^Throwable v)
              (Retried. ^Object v))))))

(defn reset!
  "Short-circuits the execution and restarts the retry cycle with the new seq
  of `delays`."
  [delays]
  (throw (Reset. delays)))

(defn marker?
  "True if the value `v` is an abort/retry/reset marker.

  Useful in case you are catching an `Throwable/Error` in the retriable code
  block and want to make sure you only act on a non-marker value, e.g.:

  ```
  (with-retries ...
    ...
    (catch Throwable t
      (when-not (marker? t)
        (do-something-on-failure)
      (throw t))))
  ```"
  [v]
  (or (identical? v abort)
      (identical? v retry)
      (instance? Reset v)
      (instance? Aborted v)
      (instance? Retried v)
      (instance? RetriesExceeded v)))

(defn- unwrap-marker [v]
  (cond (instance? Aborted v) (or (.value ^Aborted v) (.getCause ^Throwable v))
        (instance? Retried v) (or (.value ^Retried v) (.getCause ^Throwable v))
        :else v))

(defn- retries-exceeded [retried result]
  (let [result' (unwrap-marker result)]
    (if (instance? Throwable result')
      (doto (RetriesExceeded. retried)
        (.initCause result'))
      (RetriesExceeded. retried result'))))

(defn unwrap
  "Unwraps a value/cause of a `RetriesExceeded`/`Aborted`/`Retried`."
  [v]
  (cond (instance? RetriesExceeded v)
        (or (.value ^RetriesExceeded v) (.getCause ^Throwable v))
        :else (unwrap-marker v)))

(def ^:private current-thread-executor
  (reify java.util.concurrent.Executor
    (execute [_ r]
      (r))))

(def ^{:dynamic true
       :doc "The result/exception which happened during the last
 execution. `:manifail/none` during the first execution of if the
 execution was reset."}
  *last-result* nil)

(def ^{:dynamic true
       :doc "The retry count of the current execution.
Zero during the first one or after a reset."}
  *retry-count* nil)

(def ^{:dynamic true
       :doc "The number of milliseconds spent on executions and retries up to
the current execution."}
  *elapsed-ms* nil)

(defn with-retries*
  "Executes the given function `f` at least once and then as long as retry
  seq `delays` isn't drained and execution is deemed retriable.

  A retriable execution is one which fails due to a (non-abort) exception or
  returns a special `retry` value.

  `delays` is a seq producing numbers of milliseconds to wait after each
  execution. Zero means no delay, a negative delay means that execution has to
  stop.

  If you want to specify an executor which should be used to run `f`, use
  `manifold.executor/with-executor`, e.g.:

  ```
  (manifold.executor/with-executor my-executor
    (with-retries* (retries 5)
      call-service))
  ```"
  [delays f]
  (let [reset-delays (fn [^Reset r] (.retryDelays r))
        ex (or (ex/executor) current-thread-executor)
        started (System/nanoTime)]
    (d/loop [retried 0, last-result ::none, delays' delays]
      (-> (d/future-with ex
            (binding [*last-result* last-result
                      *retry-count* retried
                      *elapsed-ms* (/ (- (System/nanoTime) started) 1e6)]
              (f)))
          (d/catch' identity)
          (d/chain'
            (fn [result]
              (cond (instance? Reset result) (d/recur 0 ::none (reset-delays result))
                    (instance? Aborted result) (throw result)
                    (identical? result abort) (abort!)

                    (or (instance? Throwable result)
                        (identical? result retry))
                    (-> (let [d (first delays')]
                          (when (or (nil? d) (< d 0))
                            (throw (retries-exceeded retried result)))
                          (-> (d/deferred ex)
                              (d/timeout! d ::run)))
                        (d/chain' (fn [_] (d/recur (inc retried)
                                                   (unwrap-marker result)
                                                   (rest delays')))))
                    :else result)))))))

(defmacro with-retries
  "Macro wrapper over `with-retries*`.

  See docs for `with-retries*`."
  [delays & body]
  `(with-retries* ~delays (fn [] ~@body)))


examples/src/example.clj
(ns example
  (:require [java-time :as j]
            [manifold.deferred :as d]
            [manifail :as f]))

(defn failing-service []
  (let [r (rand-int 20)]
    (cond
      (> r 18) {:status :abort, :result r}
      (> r 12) {:status :failure, :result r}
      (> r 8) (throw (Exception. "Failed!"))
      (> r 3) {:status :unauthorized}
      (> r 2) (throw (Exception. "Fine!"))
      :else {:status :ok, :result r})))

(defn authorize []
  (println "Authorized!"))

(defn call-with-retries []
  (let [retry-delays #(f/delay (f/retries 5) 50 {:jitter-factor 0.5})]
    (-> (f/with-retries (retry-delays)
          (println "Execution n." (inc f/*retry-count*) "after" f/*elapsed-ms* "ms...")
          (when f/*last-result*
            (println "Last result:" f/*last-result*))
          (try (let [{:keys [status result]} (failing-service)]
                 (println "Got status=" status ", result=" result)
                 (when (= status :abort)
                   (println "-- on before abort")
                   (f/abort! result))
                 (when (= status :failure)
                   (println "-- on before retry 1")
                   (f/retry! result))
                 (when (= status :unauthorized)
                   (println "-- on recoverable error")
                   (authorize)
                   (println "-- on before reset")
                   (f/reset! (retry-delays)))
                 result)
               (catch Exception e
                 (if (= (.getMessage e) "Fine!")
                   :ok
                   (throw e)))
               (catch Throwable t
                 (when-not (f/marker? t)
                   (println "-- on before retry 2"))
                 (throw t))))
        (d/chain #(println "-- on complete" %))
        (d/catch manifail.Aborted #(println "-- on after abort" (f/unwrap %)))
        (d/catch manifail.RetriesExceeded #(println "-- on retries exceeded" (f/unwrap %)))
        (d/catch #(println "-- on failure" %)))))
